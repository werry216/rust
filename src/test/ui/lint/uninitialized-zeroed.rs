// ignore-tidy-linelength
// This test checks that calling `mem::{uninitialized,zeroed}` with certain types results
// in a lint.

#![feature(never_type)]
#![allow(deprecated)]
#![deny(invalid_value)]

use std::mem;

fn main() {
    unsafe {
        let _val: ! = mem::zeroed(); //~ ERROR: does not permit zero-initialization
        let _val: ! = mem::uninitialized(); //~ ERROR: does not permit being left uninitialized

        let _val: &'static i32 = mem::zeroed(); //~ ERROR: does not permit zero-initialization
        let _val: &'static i32 = mem::uninitialized(); //~ ERROR: does not permit being left uninitialized

        let _val: fn() = mem::zeroed(); //~ ERROR: does not permit zero-initialization
        let _val: fn() = mem::uninitialized(); //~ ERROR: does not permit being left uninitialized

        // Some types that should work just fine.
        let _val: Option<&'static i32> = mem::zeroed();
        let _val: Option<fn()> = mem::zeroed();
        let _val: bool = mem::zeroed();
        let _val: i32 = mem::zeroed();
    }
}
