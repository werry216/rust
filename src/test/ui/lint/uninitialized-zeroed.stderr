error: the type `!` does not permit zero-initialization
  --> $DIR/uninitialized-zeroed.rs:15:23
   |
LL |         let _val: ! = mem::zeroed();
   |                       ^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/uninitialized-zeroed.rs:7:9
   |
LL | #![deny(invalid_value)]
   |         ^^^^^^^^^^^^^

error: the type `!` does not permit being left uninitialized
  --> $DIR/uninitialized-zeroed.rs:16:23
   |
LL |         let _val: ! = mem::uninitialized();
   |                       ^^^^^^^^^^^^^^^^^^^^

error: the type `&'static i32` does not permit zero-initialization
  --> $DIR/uninitialized-zeroed.rs:21:34
   |
LL |         let _val: &'static i32 = mem::zeroed();
   |                                  ^^^^^^^^^^^^^

error: the type `&'static i32` does not permit being left uninitialized
  --> $DIR/uninitialized-zeroed.rs:22:34
   |
LL |         let _val: &'static i32 = mem::uninitialized();
   |                                  ^^^^^^^^^^^^^^^^^^^^

error: the type `fn()` does not permit zero-initialization
  --> $DIR/uninitialized-zeroed.rs:24:26
   |
LL |         let _val: fn() = mem::zeroed();
   |                          ^^^^^^^^^^^^^

error: the type `fn()` does not permit being left uninitialized
  --> $DIR/uninitialized-zeroed.rs:25:26
   |
LL |         let _val: fn() = mem::uninitialized();
   |                          ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

